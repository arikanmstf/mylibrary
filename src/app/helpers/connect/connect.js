// @flow
import React, { ComponentType } from 'react';

const connect = (mapUiToProps: Function) => (
  (WrappedComponent: ComponentType<*>) => {
    const uiComponents = mapUiToProps();
    return ({
      location, navigation, match, ...other
    }: Object) => (
      <WrappedComponent
        {...uiComponents}
        {...other}
        navigation={navigation}
        location={location || { pathname: navigation ? navigation.state.routeName : null }}
        match={match || { params: { id: navigation ? navigation.getParam('id') : undefined } }}
      />
    );
  }
);

export default connect;
