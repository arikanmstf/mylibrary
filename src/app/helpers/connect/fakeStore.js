class Store {
  constructor(props) {
    this.props = props;
  }

  toJS() {
    return this.props;
  }
}

class FakeStore {
  constructor(props) {
    this.props = props;
  }

  getState() {
    return new Store(this.props);
  }

  subscribe() {}
}

export const createFakeStore = (data) => {
  return new FakeStore(data);
};

export const dispatch = (action) => (action && typeof action === 'function' ? action() : undefined);
