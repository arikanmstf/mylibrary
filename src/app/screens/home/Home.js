/**
 * Screen Template By => create-module script
 * @version 1.0.0
 *
 */

// @flow
import * as React from 'react';
import logger from 'helpers/logger';
import type { HomeProps } from './HomeTypes';

// eslint-disable-next-line
class Home extends React.Component<HomeProps> {
  componentDidMount() {
    const {
      fetchPublications,
      fetchLists,
      cards,
      rows,
    } = this.props;

    if (!cards) {
      fetchPublications();
    }

    if (!rows) {
      fetchLists();
    }
  }

  render() {
    const {
      Screen,
      Header,
      Page,
      CardList,
    } = this.props;
    logger.log('render: Home');

    return (
      <Screen>
        <Header />
        <Page>
          <CardList />
        </Page>
      </Screen>
    );
  }
}

export default Home;
