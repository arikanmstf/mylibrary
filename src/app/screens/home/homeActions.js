/**
 * Actions Template By => create-module script
 * @version 1.0.0
 *
 */

// @flow
import { fetchAndUpdateCards } from 'modules/card/actions';
import { toggleFavorite, toggleRead } from 'modules/publication/actions';

import type { ImmutableState } from 'store/StateTypes';

export const mapStateToProps = (state: ImmutableState) => ({
  cards: state.toJS().card.cards,
  listType: state.toJS().card.listType,
  fetchedPublicationListType: state.toJS().card.fetchedPublicationListType,
});

export const mapDispatchToProps = {
  fetchPublications: fetchAndUpdateCards,
  toggleFavorite,
  toggleRead,
};
