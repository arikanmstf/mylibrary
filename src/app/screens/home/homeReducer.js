/**
 * Reducers Template By => create-module script
 * @version 1.0.0
 *
 */

import { handleActions } from 'redux-actions';
import {
  addCards,
  updateCards,
  updateTotalPages,
  updateCurrentPage,
  updateSearchQuery,
  updateSearchPending,
  updateRows,
} from './homeActions';

const initialState = {
  cards: null,
  totalPages: 0,
  currentPage: 1,
  searchQuery: '',
  isSearchPending: false,
  rows: null,
};

const reducer = new Map([
  [
    addCards,
    (state, action) => ({
      ...state,
      cards: action.payload ? [...state.cards, ...action.payload] : state.cards,
    }),
  ],
  [
    updateCards,
    (state, action) => ({
      ...state,
      cards: action.payload ? action.payload : initialState.cards,
    }),
  ],
  [
    updateTotalPages,
    (state, action) => ({
      ...state,
      totalPages: action.payload ? action.payload : initialState.totalPages,
    }),
  ],
  [
    updateCurrentPage,
    (state, action) => ({
      ...state,
      currentPage: action.payload ? action.payload : initialState.currentPage,
    }),
  ],
  [
    updateSearchQuery,
    (state, action) => ({
      ...state,
      searchQuery: action.payload ? action.payload : initialState.searchQuery,
    }),
  ],
  [
    updateSearchPending,
    (state, action) => ({
      ...state,
      isSearchPending: action.payload ? action.payload : initialState.isSearchPending,
    }),
  ],
  [
    updateRows,
    (state, action) => ({
      ...state,
      rows: action.payload || state.rows,
    }),
  ],
]);

export default handleActions(
  reducer,
  initialState
);
