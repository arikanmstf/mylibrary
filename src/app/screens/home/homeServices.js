// @flow

import Api, {
  PUBLICATION,
  TOGGLE_FAVORITE,
  TOGGLE_READ,
  LIST_MY,
} from 'helpers/api';
import { transformPublicationListToCardList, transformItemListToRowList } from 'helpers/data/transform';

import type { Pagination, CardItem } from 'ui/CardList/types';
import type {
  PublicationDetail,
  ToggleFavorite,
  ToggleRead,
  Item,
} from 'helpers/api/types';
import type { SubmitSearchFormRequest } from 'ui/Header/types';

export const getPublicationList = async (
  { page, search }: SubmitSearchFormRequest = { page: 0, search: '' }
): Promise<Pagination<CardItem>> => {
  const result: Pagination<PublicationDetail> = await Api.get(PUBLICATION, { page, search });
  return transformPublicationListToCardList(result);
};

export const toggleFavorite = async (id: number): Promise<ToggleFavorite> => {
  const result: ToggleFavorite = await Api.post(`${TOGGLE_FAVORITE}/${id}`);
  return result;
};

export const toggleRead = async (id: number): Promise<ToggleRead> => {
  const result: ToggleRead = await Api.post(`${TOGGLE_READ}/${id}`);
  return result;
};

export const getListList = async (): Promise<Array<Item>> => {
  const result: Array<Item> = await Api.get(LIST_MY);
  return transformItemListToRowList(result);
};
