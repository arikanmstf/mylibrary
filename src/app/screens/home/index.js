/**
 * Index Template By => create-module script
 * @version 1.0.0
 *
 */

export { default as Web } from './HomeWebContainer';
export { default as Native } from './HomeNativeContainer';
