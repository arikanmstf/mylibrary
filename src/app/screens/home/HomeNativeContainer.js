/**
 * Native Screen Template By => create-module script
 * @version 1.0.0
 *
 */

import { connect } from 'react-redux';
import { connect as connectUi } from 'helpers/connect';
import {
  Screen,
  Header,
  Page,
  CardList,
} from 'ui/native';
import HomeScreen from './Home';
import { mapStateToProps, mapDispatchToProps } from './homeActions';

const mapUiToProps = () => ({
  Screen,
  Header,
  Page,
  CardList,
});
export default connect(mapStateToProps, mapDispatchToProps)(connectUi(mapUiToProps)(HomeScreen));
