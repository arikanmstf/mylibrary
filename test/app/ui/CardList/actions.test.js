import {
  mapStateToProps,
  mapDispatchToProps,
  toggleFavorite,
  toggleRead,
} from 'ui/CardList/actions';
import * as homeActions from 'screens/home/homeActions';
import * as homeServices from 'screens/home/homeServices';
import { createFakeStore, dispatch } from 'helpers/connect/fakeStore';

const state = createFakeStore({
  home: {
    cards: [{
      isFavorite: false,
      isRead: false,
    }],
    search: 'dummySearch',
  },
  loader: { isLoaderVisible: false },
}).getState();

const getState = () => state;

describe('test/app/ui/CardList/actions.test.js', () => {
  homeActions.updateCards = jest.fn();
  homeActions.fetchAndAddCards = jest.fn();
  homeActions.fetchAndUpdateCards = jest.fn();
  homeServices.toggleFavorite = jest.fn(() => ({
    result: true,
  }));
  homeServices.toggleRead = jest.fn(() => ({
    result: true,
  }));

  beforeAll(() => {
    homeActions.updateCards.mockClear();
    homeActions.fetchAndAddCards.mockClear();
    homeActions.fetchAndUpdateCards.mockClear();
    homeServices.toggleFavorite.mockClear();
    homeServices.toggleRead.mockClear();
  });

  it('mapDispatchToProps', () => {
    expect(mapDispatchToProps).toMatchSnapshot();
  });

  it('mapStateToProps', () => {
    expect(mapStateToProps(state)).toMatchSnapshot();
  });

  it('toggleFavorite', async () => {
    await toggleFavorite(1, 0)(dispatch, getState);
    expect(homeServices.toggleFavorite).toBeCalledWith(1);
    expect(homeActions.updateCards).toBeCalledWith([{
      isFavorite: true,
      isRead: false,
    }]);
  });

  it('toggleRead', async () => {
    state.toJS().home.cards = [{
      isFavorite: false,
      isRead: false,
    }];

    await toggleRead(1, 0)(dispatch, getState);
    expect(homeServices.toggleRead).toBeCalledWith(1);
    expect(homeActions.updateCards).toBeCalledWith([{
      isFavorite: false,
      isRead: true,
    }]);
  });
});
