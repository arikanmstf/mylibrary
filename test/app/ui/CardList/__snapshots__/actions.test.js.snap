// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/app/ui/CardList/actions.test.js mapDispatchToProps 1`] = `
Object {
  "addCards": [Function],
  "fetchPublications": [Function],
}
`;

exports[`test/app/ui/CardList/actions.test.js mapStateToProps 1`] = `
Object {
  "cards": Array [
    Object {
      "isFavorite": false,
      "isRead": false,
    },
  ],
  "isLoaderVisible": undefined,
  "search": undefined,
}
`;
